File: mobile/android/app/src/main/kotlin/org/torchat/mobile/RuntimeSendEffect.kt
package org.torchat.mobile

import org.json.JSONObject

sealed class RuntimeSendEffect {
    enum class PairingKind(val wireValue: String) {
        OFFER("OFFER"),
        REJECTION("REJECTION"),
    }

    data class Message(
        val messageId: String,
        val conversationId: String,
        val recipientInstallationId: String,
        val body: String,
    ) : RuntimeSendEffect()

    data class Pairing(
        val pairingId: String,
        val recipientInstallationId: String,
        val kind: PairingKind,
        val payload: String?,
    ) : RuntimeSendEffect()

    companion object {
        fun fromJson(effect: JSONObject): RuntimeSendEffect =
            when {
                effect.has("messageId") -> Message(
                    messageId = effect.getString("messageId"),
                    conversationId = effect.getString("conversationId"),
                    recipientInstallationId = effect.getString("recipientInstallationId"),
                    body = effect.getString("body"),
                )
                effect.has("pairingId") -> Pairing(
                    pairingId = effect.getString("pairingId"),
                    recipientInstallationId = effect.getString("recipientInstallationId"),
                    kind = when (effect.getString("kind")) {
                        PairingKind.OFFER.wireValue -> PairingKind.OFFER
                        PairingKind.REJECTION.wireValue -> PairingKind.REJECTION
                        else -> error("unknown pairing send effect kind")
                    },
                    payload = if (effect.isNull("payload")) null else effect.getString("payload"),
                )
                else -> error("unknown runtime send effect")
            }
    }
}

File: packages/torchat-runtime/tests/runtime_send_effect.rs
use torchat_runtime::{
    MessageSendEffect, PairingSendEffect, PairingSendKind, RuntimeSendEffect,
};

#[test]
fn runtime_send_effect_serializes_like_existing_transport_payloads() {
    let message = RuntimeSendEffect::from(MessageSendEffect {
        message_id: "message-1".to_owned(),
        conversation_id: "conversation-1".to_owned(),
        recipient_installation_id: "installation-bob".to_owned(),
        body: "hello".to_owned(),
        reply_to: None,
    });
    let pairing = RuntimeSendEffect::from(PairingSendEffect {
        pairing_id: "pairing-1".to_owned(),
        recipient_installation_id: "installation-bob".to_owned(),
        kind: PairingSendKind::Offer,
        payload: Some("payload".to_owned()),
    });

    let message_json = serde_json::to_value(&message).expect("message send effect serializes");
    let pairing_json = serde_json::to_value(&pairing).expect("pairing send effect serializes");

    assert_eq!(message_json["messageId"], "message-1");
    assert_eq!(message_json["conversationId"], "conversation-1");
    assert_eq!(message_json["recipientInstallationId"], "installation-bob");
    assert_eq!(message_json["body"], "hello");

    assert_eq!(pairing_json["pairingId"], "pairing-1");
    assert_eq!(pairing_json["recipientInstallationId"], "installation-bob");
    assert_eq!(pairing_json["kind"], "OFFER");
    assert_eq!(pairing_json["payload"], "payload");
}

#[test]
fn runtime_send_effect_round_trips_from_message_and_pairing_payloads() {
    let message = serde_json::json!({
        "messageId": "message-1",
        "conversationId": "conversation-1",
        "recipientInstallationId": "installation-bob",
        "body": "hello"
    });
    let pairing = serde_json::json!({
        "pairingId": "pairing-1",
        "recipientInstallationId": "installation-bob",
        "kind": "REJECTION",
        "payload": null
    });

    let message_effect: RuntimeSendEffect =
        serde_json::from_value(message).expect("message effect parses");
    let pairing_effect: RuntimeSendEffect =
        serde_json::from_value(pairing).expect("pairing effect parses");

    assert_eq!(
        message_effect.recipient_installation_id(),
        "installation-bob"
    );
    assert_eq!(
        pairing_effect.recipient_installation_id(),
        "installation-bob"
    );
}
